src/externalized/strategic/CreatureLairModel.cc

```diff
@@ -0,0 +1,259 @@
+{
+	unsigned int totalChance = 0;
+	for (auto sec : attackSectors)
+	{
+		totalChance += sec.chance;
+	}
+
+	int random = Random(totalChance);
+	for (auto& sec : attackSectors)
+	{
+		random -= sec.chance;
+		if (random < 0)
+		{
+			return &sec;
+		}
+	}
+
+	SLOGA("Some coding error with the random selection logic. We should never reach here");
+	return NULL;
+}
+
+const CreatureAttackSector* CreatureLairModel::getTownAttackDetails(uint8_t sectorId) const
+{
+	for (auto& sec : attackSectors)
+	{
+		if (sec.sectorId == sectorId) 
+		{
+			return &sec;
+		}
+	}
+	return NULL;
+}
+
+CreatureLairModel* CreatureLairModel::deserialize(const rapidjson::Value& json)
+{
+	JsonObjectReader obj(json);
+
+	auto entrance = json["entranceSector"].GetArray();
+	auto sectorString = entrance[0].GetString();
+	if (!IS_VALID_SECTOR_SHORT_STRING(sectorString))
+	{
+		throw new std::runtime_error("Entrance sector is invalid");
+	}
+	uint8_t entranceSector = SECTOR_FROM_SECTOR_SHORT_STRING(sectorString);
+	uint8_t entranceSectorLevel = entrance[1].GetUint();
+
+	auto warpExit = JsonObjectReader(json["warpExit"]);
+	sectorString = warpExit.GetString("sector");
+	if (!IS_VALID_SECTOR_SHORT_STRING(sectorString))
+	{
+		throw new std::runtime_error("Warp exit sector is invalid");
+	}
+	uint8_t warpExitSector = SECTOR_FROM_SECTOR_SHORT_STRING(sectorString);
+	uint16_t warpExitGridNo = warpExit.GetUInt("gridNo");
+
+	return new CreatureLairModel(
+		obj.GetUInt("lairId"),
+		obj.GetUInt("associatedMineId"),
+		entranceSector, entranceSectorLevel,
+		readLairSectors(json["sectors"]),
+		readAttackSectors(json["attackSectors"]),
+		warpExitSector, warpExitGridNo
+	);
+}
+
+void CreatureLairModel::validateData(const std::vector<const CreatureLairModel*> lairs, const std::vector<const UndergroundSectorModel*> ugSectors, const uint8_t numMines)
+{
+	std::set<uint8_t> distinctLairIds;
+	for (auto lair : lairs)
+	{
+		// checkd for valid lairId, 0 is reserved
+		if (lair->lairId < 1)
+		{
+			SLOGE(ST::format("lairID {} is invalid. Must be greater than 0"));
+			throw std::runtime_error("");
+		}
+
+		// make sure we do not define the same lairId more than once
+		if (distinctLairIds.find(lair->lairId) != distinctLairIds.end())
+		{
+			SLOGE(ST::format("lairID {} is already defined before"));
+			throw std::runtime_error("");
+		}
+		distinctLairIds.insert(lair->lairId);
+
+		// if the mineId is valid
+		if (lair->associatedMineId > numMines)
+		{
+			SLOGE(ST::format("Invalid mineId {}", lair->associatedMineId));
+			throw std::runtime_error("");
+		}
+		
+		// The first lair sector in list should be QUEEN_LAIR
+		if (lair->lairSectors.size() < 1 || lair->lairSectors[0].habitatType != QUEEN_LAIR)
+		{
+			SLOGW(ST::format("The list of lair sectors should be non-empty and begins with the QUEEN_LAIR. Lair ID: {}", lair->lairId));
+		}
+
+		// all lair sectors should be adjacent and defined as underground sector
```

Comment:
From the code it looks like it's not enough that they're adjacent, but also consecutive, iow non-branching. Worth noting in the json field description.

Reply:
Actually that is a "should" and log WARN. It might work to have it non-adjacent or branching. But it probably won't make sense in terms of gameplay and story, that effectively the creatures teleports across the map.

I haven't tested if that introduce other problems. 

Adding this to field description anyways.